### PR TITLE
Add to the engine server an api method that leverages risklib to validate nrml

### PR DESCRIPTION
**Pull request by ptormene**
I've also implemented 3 tests:
* attempting to validate a valid nrml
* attempting to validate an invalid nrml
* calling the api without providing the required parameter (the xml text to be validated)

This api can be used by the risk input preparation toolkit, to check if the xml produced by the toolkit is a valid nrml.

To use it, you make a POST request to the engine server, `host/v1/valid/` with the mandatory parameter `xml_text` containing the xml to be validated. The response contains a json object with the following attributes:
* 'valid': a boolean indicating if the provided text is a valid nrml
* 'error_msg': the validation error message, if any error was found
* 'error_line': line of the given xml where the error was found

The return status code is 200 both if the NRML is valid or invalid. If the mandatory parameter xml_text is not provided, a 400 code is returned.

Tests are green: https://ci.openquake.org/job/zdevel_oq-engine/1554

### CODE REVIEW

File: openquake/server/tests/data/vulnerability_model_invalid.xml [hunk 1]
```diff
@@ -0,0 +1,59 @@
...
+            </discreteVulnerability>
+			<discreteVulnerability vulnerabilityFunctionID="W-SLFB-1" probabilisticDistribution="LN">
+                   <lossRatio>0.00028498	0.0023533	0.0090565	0.022609	0.044559	0.076123	0.11779	0.16957	0.23059	0.29953	0.37414	0.45194	0.53011	0.60612	0.67766	0.74264	0.79988	0.84854	0.88861	0.92046	0.94495	0.96308	0.97604	0.98502	0.99095	0.99472	0.99706	0.99838	0.99914	0.99957	1	1	1	1	1	1	1	1	1	1	1	1	1	1	1	1	1	1	1	1</lossRatio>
+                  <coefficientsVariation>0.00037997	0.0028769	0.01069	0.026463	0.050895	0.084018	0.12433	0.1688	0.21288	0.25159	0.28065	0.29646	0.2979	0.2854	0.26116	0.22849	0.1913	0.15358	0.11837	0.087579	0.062307	0.042706	0.02806	0.017718	0.010778	0.0062203	0.0035689	0.0020238	0.0010342	0.00052825	0	0	0	0	0	0	0	0	0	0	0	0	0	0	0	0	0	0	0	0</coefficientsVariation>
+            </discreteVulnerability>
+        </discreteVulnerabilitySet>
+    </vulnerabilityModel>
+</nrml>
```
Comment (micheles) [inline openquake/server/tests/data/vulnerability_model_invalid.xml]: Please reduce the size of this file to maximum 10 lines. Also, try to use negative coefficients to make this invalid.

File: openquake/server/views.py [hunk 2]
```diff
@@ -145,6 +147,39 @@ def get_engine_version(request):
...
+        nrml.read(xml_file)
+    except Exception as exc:
+        response_data['validation_errors'] = str(exc).splitlines()
+        response_data['valid'] = False
+    else:
+        response_data['validation_errors'] = []
+        response_data['valid'] = True
+    return HttpResponse(
+        content=json.dumps(response_data), content_type=JSON)
+
```
Comment (micheles) [inline openquake/server/views.py]: I would return the line number as a number in the JSON structure.

File: openquake/server/tests/data/vulnerability_model_invalid.xml (hunk 1 above)
Comment (ptormene) [inline openquake/server/tests/data/vulnerability_model_invalid.xml]: Done

File: openquake/server/views.py (hunk 2 above)
Comment (ptormene) [inline openquake/server/views.py]: Done

File: openquake/server/tests/functional_test.py [hunk 3]
```diff
@@ -148,4 +154,44 @@ def test_err_3(self):
...
+    def test_validate_nrml_valid(self):
+        valid_file = os.path.join(self.datadir, 'vulnerability_model.xml')
+        with open(valid_file, 'rb') as vf:
+            valid_content = vf.read()
+        data = dict(xml_text=valid_content)
+        resp = self.post_nrml(data)
+        assert resp.status_code == 200, resp
+        resp_text_dict = json.loads(resp.text)
+        assert resp_text_dict['valid'], resp_text_dict
+        assert resp_text_dict['error_msg'] is None, resp_text_dict
```
Comment (micheles) [inline openquake/server/tests/functional_test.py]: You can user self.assertIsNone

File: openquake/server/tests/functional_test.py [hunk 4]
```diff
@@ -148,4 +154,44 @@ def test_err_3(self):
...
+        data = dict(xml_text=invalid_content)
+        resp = self.post_nrml(data)
+        assert resp.status_code == 200, resp
+        resp_text_dict = json.loads(resp.text)
+        assert not resp_text_dict['valid'], resp_text_dict
+        expected_error_line = 7
+        expected_error_msg = (u'Could not convert lossRatio->positivefloats:'
+                              ' float -0.018800826 < 0')
+        assert resp_text_dict['error_msg'] == \
+            expected_error_msg, resp_text_dict['error_msg']
```
Comment (micheles) [inline openquake/server/tests/functional_test.py]: Use `self.assertEqual`

File: openquake/server/tests/functional_test.py [hunk 5]
```diff
@@ -148,4 +154,44 @@ def test_err_3(self):
...
+        assert resp_text_dict['error_line'] == \
+            expected_error_line, resp_text_dict['error_line']
+
+    def test_validate_nrml_missing_parameter(self):
+        # passing a wrong parameter, instead of the required 'xml_text'
+        data = dict(foo="bar")
+        resp = self.post_nrml(data)
+        assert resp.status_code == 400, resp
+        assert \
+            resp.text == 'Please provide the "xml_text" parameter', resp.text
```
Comment (micheles) [inline openquake/server/tests/functional_test.py]: Use `self.assertEqual`

File: openquake/server/views.py [hunk 6]
```diff
@@ -145,6 +148,46 @@ def get_engine_version(request):
...
+    xml_text = request.POST.get('xml_text')
+    if not xml_text:
+        return HttpResponseBadRequest(
+            'Please provide the "xml_text" parameter')
+    xml_file = writetmp(xml_text, suffix='.xml')
+    response_data = {}
+    try:
+        nrml.read(xml_file)
+    except Exception as exc:
+        error_msg = str(exc).split(', line')[0]
```
Comment (micheles) [inline openquake/server/views.py]: Inside `exc` there could be an UTF8-encoded string, so `str(exc)` may fail. Do some experiments and find the right combination to manage that case.

Comment (ptormene) [inline openquake/server/views.py]: Could the same happen also inside `run_calc`? If so, I should do the same also there.

File: openquake/server/tests/functional_test.py (hunk 3 above)
Comment (ptormene) [inline openquake/server/tests/functional_test.py]: Done

File: openquake/server/tests/functional_test.py (hunk 4 above)
Comment (ptormene) [inline openquake/server/tests/functional_test.py]: Done

File: openquake/server/tests/functional_test.py (hunk 5 above)
Comment (ptormene) [inline openquake/server/tests/functional_test.py]: Done

File: openquake/server/views.py (hunk 6 above)
Comment (ptormene) [inline openquake/server/views.py]: I've fixed it also inside `run_calc`.

File: openquake/server/views.py [hunk 7]
```diff
@@ -145,6 +148,52 @@ def get_engine_version(request):
...
+            'Please provide the "xml_text" parameter')
+    xml_file = writetmp(xml_text, suffix='.xml')
+    response_data = {}
+    try:
+        nrml.read(xml_file)
+    except Exception as exc:
+        # get the exception message
+        exc_msg = exc.args[0]
+        # if it is not a simple string, let's assume it's UTF8 eoncoded
+        if not isinstance(exc_msg, bytes):
```
Comment (micheles) [inline openquake/server/views.py]: the commend should be `if it is a unicode object, encode it as a UTF-8 bytestring`

File: openquake/server/views.py [hunk 8]
```diff
@@ -145,6 +148,52 @@ def get_engine_version(request):
...
+            exc_msg = exc_msg.encode('utf8')
+        error_msg = exc_msg.split(', line')[0]
```
Comment (micheles) [inline openquake/server/views.py]: I would use the logic

```python
if isinstance(exc_msg, bytes):
    exc_msg = exc_msg.decode('utf-8')   # make it a unicode object
else:
    assert isinstance(exc_msg, unicode), exc_msg

```

File: openquake/server/views.py [hunk 9]
```diff
@@ -145,6 +148,52 @@ def get_engine_version(request):
...
+        error_line = int(
+            re.search(r'line \d+', exc_msg).group(0).split()[1])
+        response_data['error_msg'] = error_msg
```
Comment (micheles) [inline openquake/server/views.py]: I would be more careful here. If the error message does not contain a line number, return '-1' as error_line.

Comment (ptormene) [inline openquake/server/views.py]: Done

File: openquake/server/views.py (hunk 8 above)
Comment (ptormene) [inline openquake/server/views.py]: Done

File: engine_api.md [hunk 10]
```diff
@@ -165,3 +165,19 @@ Parameters:
     * hazard_result: the hazard results ID upon which to run the risk calculation; specify this or hazard_calc (only for risk calculations)
 
 Response: Redirects to [/v1/calc/:calc_id](#get-v1calchazardcalc_id), where `calc_id` is the ID of the newly created calculation.
+
+
+#### POST /v1/valid/
+
+Leverage oq-risklib to check if a given xml text is a valid nrml.
+
```
Comment (micheles) [inline engine_api.md]: Use NRML instead of nrml

File: engine_api.md [hunk 11]
```diff
@@ -165,3 +165,19 @@ Parameters:
...
+Parameters:
+
+    * xml_text: the text of the xml to be validated as nrml
+
```
Comment (micheles) [inline engine_api.md]: Idem

File: engine_api.md [hunk 12]
```diff
@@ -165,3 +165,19 @@ Parameters:
...
+Response:
+
+a JSON object, containing:
+    * valid: a boolean indicating if the provided text is a valid nrml
+    * error_msg: the error message, if any error was found (None otherwise)
```
Comment (micheles) [inline engine_api.md]: Idem

File: openquake/server/tests/functional_test.py [hunk 13]
```diff
@@ -148,4 +154,56 @@ def test_err_3(self):
...
+    # tests for nrml validation
+
+    def test_validate_nrml_valid(self):
+        valid_file = os.path.join(self.datadir, 'vulnerability_model.xml')
+        with open(valid_file, 'rb') as vf:
+            valid_content = vf.read()
+        data = dict(xml_text=valid_content)
+        resp = self.post_nrml(data)
+        self.assertEqual(resp.status_code, 200)
+        resp_text_dict = json.loads(resp.text)
```
Comment (micheles) [inline openquake/server/tests/functional_test.py]: Perhaps `resp.content`? check with requests library documentation.

File: openquake/server/views.py [hunk 14]
```diff
@@ -145,6 +148,70 @@ def get_engine_version(request):
...
+@csrf_exempt
+@cross_domain_ajax
+@require_http_methods(['POST'])
+def validate_nrml(request):
+    """
+    Leverage oq-risklib to check if a given xml text is a valid nrml
+
+    :param request:
+        a `django.http.HttpRequest` object containing the mandatory
+        parameter 'xml_text': the text of the xml to be validated as nrml
```
Comment (micheles) [inline openquake/server/views.py]: Please use NRML

**Comment by micheles** [comment]
Apart from the lowercase/uppercase stylistic issue, LGTM

File: engine_api.md (hunk 11 above)
Comment (ptormene) [inline engine_api.md]: Done

File: openquake/server/views.py (hunk 14 above)
Comment (ptormene) [inline openquake/server/views.py]: Done

File: engine_api.md (hunk 12 above)
Comment (ptormene) [inline engine_api.md]: Done

File: engine_api.md (hunk 10 above)
Comment (ptormene) [inline engine_api.md]: Done

**Pull request merged**